### ISSUE TITLE
Using mapbox for tileUrl issues

### ISSUE THREAD

**Issue by krish-penumarty** [StephanGeorg/staticmaps#35]
Hi @StephanGeorg Thanks for creating an awesome package. Currently using it pretty heavily to do all sorts of work, but we are hitting limitations using OSM under the hood. Getting errors based on the hit rate we have and we finally realized we need to move to a different service. I started looking at tile servers that we would use to replace the tileUrl when making the request with a paid service.

Currently I was thinking of using MapBox api to handle this for us. Here is the sample tileURL I use for this request, but I get an empty blue screen. [Documentation link](https://docs.mapbox.com/api/maps/#static-tiles).

https://api.mapbox.com/styles/v1/mapbox/streets-v11/tiles/256/{z}/{y}/{x}

I noticed that the z, x and y arguments result in `0` after diving deeper into your library. Any idea why this wouldn't work and if you have any suggestions? If this API doesn't work do you recommend a different paid service or even hosting a tile service ourselves?

Sample map request instance:
```
StaticMaps {
  options: {
    width: 400,
    height: 400,
    tileUrl: 'https://api.mapbox.com/styles/v1/mapbox/streets-v11/tiles/256/{z}/{y}/{x}?access_token=xxx'
  },
  width: 400,
  height: 400,
  paddingX: 0,
  paddingY: 0,
  padding: [ 0, 0 ],
  tileUrl: 'https://api.mapbox.com/styles/v1/mapbox/streets-v11/tiles/256/{z}/{y}/{x}?access_token=xxx',
  tileSize: 256,
  tileRequestTimeout: undefined,
  tileRequestHeader: undefined,
  reverseY: false,
  maxZoom: undefined,
  markers: [],
  lines: [
    Polyline {
      options: [Object],
      coords: [Array],
      color: '#000000BB',
      fill: undefined,
      width: 2,
      simplify: false,
      type: 'polyline'
    },
    Polyline {
      options: [Object],
      coords: [Array],
      color: '#000000BB',
      fill: undefined,
      width: 2,
      simplify: false,
      type: 'polyline'
    },
    Polyline {
      options: [Object],
      coords: [Array],
      color: '#000000BB',
      fill: undefined,
      width: 2,
      simplify: false,
      type: 'polyline'
    },
    Polyline {
      options: [Object],
      coords: [Array],
      color: '#000000BB',
      fill: undefined,
      width: 2,
      simplify: false,
      type: 'polyline'
    },
    Polyline {
      options: [Object],
      coords: [Array],
      color: '#000000BB',
      fill: undefined,
      width: 2,
      simplify: false,
      type: 'polyline'
    }
  ],
  polygons: [],
  text: [
    Text {
      options: [Object],
      coord: [Array],
      text: "63.2'",
      color: 'transparent',
      width: '0pxpx',
      fill: 'black',
      size: 10,
      font: 'Verdana',
      rotate: -272.52922142700277,
      padding: 3
    },
    Text {
      options: [Object],
      coord: [Array],
      text: "42.5'",
      color: 'transparent',
      width: '0pxpx',
      fill: 'black',
      size: 10,
      font: 'Verdana',
      rotate: -279.214017546298,
      padding: 3
    },
    Text {
      options: [Object],
      coord: [Array],
      text: "208.3'",
      color: 'transparent',
      width: '0pxpx',
      fill: 'black',
      size: 10,
      font: 'Verdana',
      rotate: 359.5756566386437,
      padding: 3
    },
    Text {
      options: [Object],
      coord: [Array],
      text: "104.9'",
      color: 'transparent',
      width: '0pxpx',
      fill: 'black',
      size: 10,
      font: 'Verdana',
      rotate: 270.05790081503244,
      padding: 3
    },
    Text {
      options: [Object],
      coord: [Array],
      text: "198.8'",
      color: 'transparent',
      width: '0pxpx',
      fill: 'black',
      size: 10,
      font: 'Verdana',
      rotate: 359.5237982574305,
      padding: 3
    }
  ],
  center: [],
  centerX: 0,
  centerY: 0,
  zoom: 0
}
```

**Comment by StephanGeorg** [comment]
I think you you mixed something. It should be ```{z}/{x}/{y}``` not ```{z}/{y}/{x} ```

**Comment by krish-penumarty** [comment]
That was it! I think I stared at this for too long and this went over my head. Thanks for checking on this and helping me resolve this! Really appreciate it.

**Comment by krish-penumarty** [comment]
Hey @StephanGeorg - Reopening this issue. Had a question about the implementation of how many tiles are fetched. I'm noticing that while using Mapbox our requests have skyrocketed to around 40k while on our end it has been 6.5k requests to static maps. I looked at the source code and it shows that there is a loop that essentially makes a list of requests to that tileUrl.

I assumed it was a single request to get the image, but it seems like you need to fetch multiple tiles to build the background map image. Could you explain how this works and what are some ways to limit the number of requests? (I was going to set a static height and width to see if that reduces number of tiles)

**Comment by StephanGeorg** [comment]
Sure. Tiles have a fixed width and height and representing a fixed area in the real world. This depends on the selected zoom level. To render the base map I need to download all tiles which are part of the base map even if I need only a small part of that tile. Than I crop the needed part of that tile and compose it to the new base map.

**Comment by krish-penumarty** [comment]
This makes a ton of sense and thank you for the super quick response. Is there anyway I can set the baseImage to be smaller so we don't request as many tile requests?

**Comment by StephanGeorg** [comment]
It is very difficult to say how requests can be reduced. The size of an image depends on the layers used (points, polylines, ...), which define the bounding box, which in turn defines the zoom level. 

> around 40k while on our end it has been 6.5k requests to static maps.

Do you render your maps on the fly?

Maybe you should cache the tiles and reuse them for your maps. This does only work if your maps are all in the same area.

**Comment by krish-penumarty** [comment]
I see! We do render maps on the fly and can be designated anywhere in the country. Unfortunately we are unable to cache tiles since they are against Mapbox user agreement. We are going to look at a few different solutions that can possibly work for us. Thanks for all your help @StephanGeorg

**Comment by krish-penumarty** [comment]
Just wanted to provide an update for a solution I ended up doing to reduce the number of tile requests. Mapbox by default tries to render tiles in 512px, so by switching over to using 512px from mapbox and setting the tileSize to 512px I reduced our calls by 30%. This is for anyone else who was looking at reducing calls. 

I also added a logging framework for a forked version to essentially capture number of requests that were being made to a remote tileUrl to generate the static tile which we were able to start capturing detailed metrics on.
